### PR TITLE
Add support for the DirectML Backend

### PR DESCRIPTION
**Pull request by smk2007**
Add support for the DirectML Backend

Summary:
This PR implements the necessary hooks/stubs/enums/etc for complete DirectML Backend integration. The actual backend implementation is currently private.

Direct Machine Learning (DirectML) is a low-level API for machine learning. DirectML is supported by all DirectX 12-compatible hardware and is layered on top of the Compute Driver Model which is the common abstraction layer to provide consistent, reliable, and secure ML execution across all Windows hardware devices and vendors.

We have been working on this at Microsoft for the last few months, and are finally ready to contribute the PyTorch core changes upstream (nothing major or exciting, just the usual boilerplate for adding new backends).

The DirectML Backend will allow us to ferry torch ops into granular DirectML powered kernels that execute against any DX12-capacble GPU device.

[RFC-0028: PyTorch-DirectML by Adele101 · Pull Request #46 · pytorch/rfcs (github.com)](https://github.com/pytorch/rfcs/pull/46)

[PyTorch-DirectML RFC · Issue #84721 · pytorch/pytorch (github.com)](https://github.com/pytorch/pytorch/issues/84721)

cc @EikanWang @jgong5 @wenzhe-nrv @sanchitintel

### CODE REVIEW

File: aten/src/ATen/detail/DMLHooksInterface.h [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  "The 'torch_dml' module is provided by "
+  "(https://github.com/microsoft/directml).";
+
+// NB: Class must live in `at` due to limitations of Registry.h.
+namespace at {
+
+struct TORCH_API DMLHooksInterface {
+  // This should never actually be implemented, but it is used to
+  // squelch -Werror=non-virtual-dtor
+  virtual ~DMLHooksInterface() {}
```
Comment (malfet) [inline aten/src/ATen/detail/DMLHooksInterface.h:17]: Nit
```suggestion
  virtual ~DMLHooksInterface() = default;
```

File: c10/core/DeviceType.h [hunk 2]
```diff
@@ -50,15 +50,18 @@ enum class DeviceType : int8_t {
   Lazy = 17, // Lazy Tensors
   IPU = 18, // Graphcore IPU
   PrivateUse1 = 19, // PrivateUse1 device
+  DML = 20, // DirectML / Microsoft
```
Comment (malfet) [inline c10/core/DeviceType.h:53]: Hmm, should PrivateUse1 always be the last one in the list? (not sure to be frank..)

File: caffe2/proto/caffe2.proto [hunk 3]
```diff
@@ -221,8 +221,9 @@ enum DeviceTypeProto {
   PROTO_ORT = 8; // ONNX Runtime
   PROTO_XLA = 9; // XLA / TPU
   PROTO_MPS = 10; // MPS
+  PROTO_DML = 11; // DirectML
   // Change the following number if you add more devices in the code.
-  PROTO_COMPILE_TIME_MAX_DEVICE_TYPES = 11;
+  PROTO_COMPILE_TIME_MAX_DEVICE_TYPES = 12;
```
Comment (malfet) [inline caffe2/proto/caffe2.proto:226]: Is DirectML support in Caffe2 really planned?

**Review by malfet — COMMENTED** [review]
Hmm, does `third_party/dml` really needs to be checked into the core repository or can it be hosted in some other repo? Regardless, why not split it into two PRs: one that adds new DispatchKey/Hooks and another that adds `third_party/dml/` boilerplate

**Comment by albanD** [comment]
By the way, we also have some ongoing efforts to allow doing this out of tree: https://github.com/bdhirsh/pytorch_open_registration_example
Did you consider this approach? And if so, what were the limitations? We would be very happy to work with you to enable full out of tree backend as it is more sustainable for us in the long term.

**Comment by smk2007** [comment]
> Hmm, does `third_party/dml` really needs to be checked into the core repository or can it be hosted in some other repo? Regardless, why not split it into two PRs: one that adds new DispatchKey/Hooks and another that adds `third_party/dml/` boilerplate

@malfet It should not be there. It is a relic from when we built the DirectML plugin in-tree and now obsolete. Removing.

**Comment by smk2007** [comment]
> By the way, we also have some ongoing efforts to allow doing this out of tree: https://github.com/bdhirsh/pytorch_open_registration_example Did you consider this approach? And if so, what were the limitations? We would be very happy to work with you to enable full out of tree backend as it is more sustainable for us in the long term.

@albanD We more or less follow the same pattern. It sure would have been handy to have this repo to reference when we began the out-of-tree plugin! However, we did not want to show up as PrivateUse1, and instead wanted a named backend to be clear to users that they are using DML. The rest of the code is just boilerplate i threw together from various PRs so the DML shows up as a named backend, but remains out-of-tree.

**Comment by chauhang** [comment]
@smk2007 Thanks for this PR. Can you please share the binaries for the DirectML plugin and add steps for CI integration and basic end to end tests with the binaries?

File: c10/core/DeviceType.h (hunk 2 above)
Comment (ezyang) [reply c10/core/DeviceType.h:53]: For this list it's OK for it not to be last. In BackendComponent, though, it's important for the private use keys to be last.

**Comment by smk2007** [comment]
> We more or less follow the same pattern. It sure would have been handy to have this repo to reference when we began the out-of-tree plugin! However, we did not want to show

Hi @albanD,
We took a look into using PrivateUse1 for the DirectML backend, and ran into some issues with the PrivateUse1 backend not supporting opaque tensors. I identified a few issues which were blocking DirectML (since DML tensors are opaque), and created this PR:
https://github.com/pytorch/pytorch/pull/86557

**Comment by EwoutH** [comment]
This looks interesting! What’s needed to move this forward?

**Comment by malfet** [comment]
Closing this one, as DirectML were implemented using PrivateUse as suggested in https://github.com/pytorch/pytorch/pull/85040#issuecomment-1254626017  see https://learn.microsoft.com/en-us/windows/ai/directml/gpu-pytorch-windows

**Comment by Looong01** [comment]
> 



> Closing this one, as DirectML were implemented using PrivateUse as suggested in [#85040 (comment)](https://github.com/pytorch/pytorch/pull/85040#issuecomment-1254626017) see https://learn.microsoft.com/en-us/windows/ai/directml/gpu-pytorch-windows

“PrivateUse”， what's that meaning? Your team cannot take part in developing of Torch-directml, just because the policy from Microsoft? Can you apply for permission?

**Comment by Looong01** [comment]
> 



> Closing this one, as DirectML were implemented using PrivateUse as suggested in [#85040 (comment)](https://github.com/pytorch/pytorch/pull/85040#issuecomment-1254626017) see https://learn.microsoft.com/en-us/windows/ai/directml/gpu-pytorch-windows

“PrivateUse”， what's that meaning? Your team cannot take part in developing of Torch-directml, just because the policy from Microsoft? Can you apply for permission?

> Closing this one, as DirectML were implemented using PrivateUse as suggested in [#85040 (comment)](https://github.com/pytorch/pytorch/pull/85040#issuecomment-1254626017) see https://learn.microsoft.com/en-us/windows/ai/directml/gpu-pytorch-windows

“PrivateUse”， what's that meaning? Your team cannot take part in developing of Torch-directml, just because the policy from Microsoft? Can you apply for permission?
We still cannot use the full operations supporting version of PyTorch on AMD GPU on Windows. Developing which is only from Microsoft team is in plight. Many ops are still not supported and those supported ops are in bad optimized with so many version updated. DirectML may be our only hope. That is too important to us. Could you help to do more effort about it.
Thank you forever!

**Comment by Looong01** [comment]
> Closing this one, as DirectML were implemented using PrivateUse as suggested in [#85040 (comment)](https://github.com/pytorch/pytorch/pull/85040#issuecomment-1254626017) see https://learn.microsoft.com/en-us/windows/ai/directml/gpu-pytorch-windows

“PrivateUse”， what's that meaning? Your team cannot take part in developing of Torch-directml, just because the policy from Microsoft? Can you apply for permission?
We still cannot use the full operations supporting version of PyTorch on AMD GPU on Windows. Developing which is only from Microsoft team is in plight. Many ops are still not supported and those supported ops are in bad optimized with so many version updated. DirectML may be our only hope. That is too important to us. Could you help to do more effort about it.
Thank you forever!

**Pull request closed without merging**